pkg/apis/eventing/v1alpha1/subscription_types.go

```diff
@@ -0,0 +1,200 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1alpha1
+
+import (
+	"encoding/json"
+
+	"github.com/knative/pkg/apis"
+	"github.com/knative/pkg/webhook"
+	"k8s.io/api/core/v1"
+	corev1 "k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:defaulter-gen=true
+
+// Subscription routes events received on a Channel to a DNS name and
+// corresponds to the subscriptions.channels.knative.dev CRD.
+type Subscription struct {
+	meta_v1.TypeMeta   `json:",inline"`
+	meta_v1.ObjectMeta `json:"metadata"`
+	Spec               SubscriptionSpec   `json:"spec"`
+	Status             SubscriptionStatus `json:"status,omitempty"`
+}
+
+// Check that Subscription can be validated, can be defaulted, and has immutable fields.
+var _ apis.Validatable = (*Subscription)(nil)
+var _ apis.Defaultable = (*Subscription)(nil)
+var _ apis.Immutable = (*Subscription)(nil)
+var _ runtime.Object = (*Subscription)(nil)
+var _ webhook.GenericCRD = (*Subscription)(nil)
+
+// SubscriptionSpec specifies the Channel for incoming events, a Call target for
+// processing those events and where to put the result of the processing. Only
+// From (where the events are coming from) is always required. You can optionally
+// only Process the events (results in no output events) by leaving out the Result.
+// You can also perform an identity transformation on the invoming events by leaving
+// out the Call and only specifying Result.
+//
+// The following are all valid specifications:
+// from --[call]--> result
+// Sink, no outgoing events:
+// from -- call
+// no-op function (identity transformation):
+// from --> result
+type SubscriptionSpec struct {
+	// TODO: Generation used to not work correctly with CRD. They were scrubbed
+	// by the APIserver (https://github.com/kubernetes/kubernetes/issues/58778)
+	// So, we add Generation here. Once the above bug gets rolled out to production
+	// clusters, remove this and use ObjectMeta.Generation instead.
+	// +optional
+	Generation int64 `json:"generation,omitempty"`
+
+	// Reference to an object that will be used to create the subscription
+	// for receiving events. The object must have spec.subscriptions
+	// list which will then be modified accordingly.
+	//
+	// This object must fulfill the Subscribable contract.
+	//
+	// You can specify only the following fields of the ObjectReference:
+	//   - Kind
+	//   - APIVersion
+	//   - Name
+	// This field is immutable
```

Comment:
Why is this field immutable?

Reply:
The idea was that since a Subscription cannot be unsubscribed from one channel and subscribed to another channel atomically that we shouldn't pretend to offer atomic mutations. There would be a (hopefully brief) period where either both channels have the subscription or neither channel has the subscription. Preventing mutation forces the user to think about the semantics they want during the migration.